### PR TITLE
Do not generate relocs to empty type dictionaries

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
If the generic dictionary is empty, there's no legitimate reason to access it. Saves 20 kB on BasicMinimalApi.

Also deleting a multimodule extensibility point that was getting in the way. We didn't use it and I don't see a future need for it either.

Cc @dotnet/ilc-contrib

### CODE REVIEW

**Review by jkotas — APPROVED** [review]
Thanks!

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EETypeNode.cs [hunk 1]
```diff
@@ -926,9 +926,18 @@ private void OutputVirtualSlots(NodeFactory factory, ref ObjectDataBuilder objDa
...
                     || isInterfaceWithAnEmptySlot)
+                {
                     objData.EmitZeroPointer();
+                }
                 else
-                    objData.EmitPointerReloc(factory.TypeGenericDictionary(declType));
+                {
+                    TypeGenericDictionaryNode dictionaryNode = factory.TypeGenericDictionary(declType);
+                    DictionaryLayoutNode layoutNode = dictionaryNode.GetDictionaryLayout(factory);
+                    if (layoutNode.HasFixedSlots && layoutNode.IsEmpty)
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EETypeNode.cs:936]: Does `layoutNode.HasFixedSlots` really needed to be in the condition? I would expect this to work for lazily computed dictionary layout too.

Comment (MichalStrehovsky) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EETypeNode.cs:936]: We need either this, or condition it on `relocsOnly`. One cannot asks about slots before we have the whole program view.

Comment (jkotas) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EETypeNode.cs:936]: `relocsOnly` would make more sense to me.

**Pull request merged**